multi_model_experiment/05_global_token_validation_experiment.py

```diff
@@ -0,0 +1,202 @@
+        for file in all_inter_year_models
+    ]
+)
+all_inter_models_df >> ply.slice_rows(15)
+
+middle_estimate = (
+    all_inter_models_df
+    >> ply.query("year_pair=='2010_0-2011_0'")
+    >> ply.pull("global_distance")
+).mean()
+middle_estimate
+
+regex_pattern = r"(\d+)_0-(\d+)_0"
+percent_diff_one_year = (
+    all_inter_models_df
+    >> ply_tdy.extract(
+        "year_pair", into=["year_1", "year_2"], regex=regex_pattern, convert=True
+    )
+    >> ply.group_by("year_1", "year_2")
+    >> ply.query("year_2-year_1==1")
+    >> ply.define(pct_diff="mean(global_distance)/middle_estimate - 1")
+    >> ply.select("year_1", "year_2", "pct_diff")
+    >> ply.ungroup()
+    >> ply.distinct()
+    >> ply_tdy.unite("year_pair", "year_1", "year_2")
+)
+percent_diff_one_year
+
+# # Does distance increase across the years?
+
+all_distance_df = (
+    full_token_set_df
+    >> ply.arrange("tok")
+    >> ply.group_by("year_2")
+    >> ply.define(
+        avg_global_distance="mean(original_global_distance)",
+        avg_global_distance_qst="mean(global_distance_qst)",
+        avg_ratio_metric="mean(ratio_metric)",
+    )
+    >> ply.select(
+        "avg_global_distance", "avg_global_distance_qst", "avg_ratio_metric", "year_1"
+    )
+    >> ply.distinct()
+    >> ply.ungroup()
+    >> ply.query("year_2-year_1==1")
+    >> ply_tdy.unite("year_pair", "year_1", "year_2")
+)
+all_distance_df >> ply.slice_rows(15)
+
+global_qst_comparison = all_distance_df >> ply_tdy.gather(
+    "metric",
+    "distance_value",
+    ply.select(
+        "avg_global_distance_qst",
+        "avg_global_distance",
+        "avg_ratio_metric",
+    ),
+)
+global_qst_comparison.year_pair = pd.Categorical(
+    global_qst_comparison.year_pair.tolist()
+)
+global_qst_comparison.head(10)
+
+middle_map = dict(
+    global_qst_comparison
+    >> ply.query("year_pair=='2010_2011'")
+    >> ply.pull(["metric", "distance_value"])
+)
+
+pct_diff_df = (
+    global_qst_comparison
+    >> ply.arrange("year_pair")
+    >> ply.define(
+        pct_diff=ply.expressions.case_when(
+            {
+                'metric=="avg_global_distance_qst"': f'distance_value/{middle_map["avg_global_distance_qst"]} - 1',
+                'metric=="avg_global_distance"': f'distance_value/{middle_map["avg_global_distance"]} - 1',
+                'metric=="avg_ratio_metric"': f'distance_value/{middle_map["avg_ratio_metric"]} - 1',
+            }
+        )
+    )
+)
+pct_diff_df
+
+(
+    p9.ggplot(
+        pct_diff_df
+        >> ply.select("-distance_value")
+        >> ply.call(
+            ".append", percent_diff_one_year >> ply.define(metric='"single_model"')
+        ),
+        p9.aes(x="year_pair", y="abs(pct_diff)", group="metric", color="metric"),
+    )
+    + p9.geom_point()
+    + p9.geom_line()
+    + p9.coord_flip()
+    + p9.theme_seaborn(style="white")
+    + p9.labs(title="Percent Difference Relative to 2010-2011")
+    + p9.scale_color_brewer("qual", palette=2)
+)
+
+g = (
```

Comment:
Here are you measuring how a word in reference to 2010-11. So you're measuring intER distance between 2010 vs 2012, 2010 vs 2013, ..etc. So you're qst metric is showing no inter changes now? As opposed to using a single model that is very variable. But you also want to detect when words are changing...so is this a control case to make to show that an expected word that shouldn't change over years doesn't?

Reply:
>Here are you measuring how a word in reference to 2010-11. So you're measuring intER distance between 2010 vs 2012, 2010 vs 2013, ..etc. So you're qst metric is showing no inter changes now? As opposed to using a single model that is very variable. But you also want to detect when words are changing...so is this a control case to make to show that an expected word that shouldn't change over years doesn't?

Close. This plot is a global picture of all word distances averaged together. The green plot shows if I use my correction metric for every model and every word averaged across the years. The orange plot shows what happens when one doesn't use a correction method or even multiple models at all. Traditionally most people just train one model per year and use that to compare the differences (assuming alignment has occurred). This plot adds some justification to why I need a correction step.